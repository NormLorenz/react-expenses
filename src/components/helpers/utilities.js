/* global google */

import moment from 'moment';

export function calculateDate(dateString, taxYear) {

  // calculate a date allowing only month and day input format
  let date = moment(dateString, ['MM-DD', 'MM-DD-YYYY']);

  // default to january 1st if improper format
  if (date.isValid() === false) {
    date = moment({ y: taxYear, m: 1, d: 1 });
  }

  // add the tax year if not explicitly declared
  else if (dateString.length <= 5) {
    date.year(taxYear);
  }

  // return something to the caller
  return { date: date.toISOString(), taxYear: date.year() };
}

export function testDate() {
  return moment({ y: 2015, m: 1, d: 1 }).toString();
}

export function sum(a, b) {
  return a + b;
}

export function convertCentsToDollars(amount) {
  let dollars = Math.floor(amount / 100);
  let cents = amount % 100;
  if (cents.toString().length === 1) { cents = '0' + cents; }
  return `${dollars}.${cents}`;
}

export function convertDollarsToCents(amount) {
  let calculatedAmount = (amount).toString().replace(/[^0-9.]/g, '');
  return Math.round(calculatedAmount * 100);
}

export function convertTextWithEllipsis(text, length) {
  let result = '';
  if (text != null && length != null) {
    result = text.substring(0, length) + '...';
  }
  return result;
}

// not enterprise ready - need to consider child nodes
export const objectToArray = (snapshot) => {
  let returnArr = [];
  snapshot.forEach((childSnapshot) => {
    let item = childSnapshot.val();
    item.key = childSnapshot.key;
    returnArr.push(item);
  });
  return returnArr;
};

// not enterprise ready
export const arrayToObject = (array) =>
  array.reduce((obj, item) => {
    obj[item.key] = item;
    return obj;
  }, {});

// export function stringifyObject (object) {
//   return <pre>{JSON.stringify(object, null, ' ')}</pre>
// }

//region mileage routines

export function calculateMileage(directions) {
  const MetersToMilesConst = 1609.334;
  let mileage = 0;

  if (directions) {
    let meters = directions.routes[0].legs
      .map(function (a) { return a.distance.value; })
      .reduce(function (a, b) { return a + b; });
    mileage = Math.round(meters / MetersToMilesConst);
  }

  return mileage;
}

export function getDirections(wayPoints, places) {
  return new Promise((resolve, reject) => {

    let legs = wayPoints
      .filter(function (a) { return a.place != null; })
      .map(function (a) {
        let place = places.find(function (b) { return a.place === b.key; });
        return { lat: place.data.latitude, lng: place.data.longitude };
      });

    if (legs.length < 2) {
      reject('must have at least two places to compute mileage');
    }

    else {
      let request = {
        origin: legs[0],
        waypoints: legs
          .slice(1, legs.length - 1)
          .map(function (leg) { return { location: leg }; }),
        destination: legs[legs.length - 1],
        travelMode: google.maps.TravelMode.DRIVING
      };

      // pass the request to the directions service
      const directionsService = new google.maps.DirectionsService();

      directionsService.route(request, (response, status) => {
        if (status === google.maps.DirectionsStatus.OK) {
          resolve(response);
        }
        else {
          reject(status);
        }
      });
    }
  });
}

//endregion

