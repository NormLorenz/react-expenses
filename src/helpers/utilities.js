
export function convertCentsToDollars(amount) {
  let dollars = Math.floor(amount / 100);
  let cents = amount % 100;
  if (cents.toString().length === 1) { cents = '0' + cents; }
  return `${dollars}.${cents}`;
}