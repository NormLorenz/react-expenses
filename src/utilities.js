// http://wesbos.com/javascript-modules/
// https://leanpub.com/understandinges6/read#leanpub-auto-encapsulating-code-with-modules

 export function convertCentsToDollars(amount) {
    let dollars = Math.floor(amount / 100);
    let cents = amount % 100;
    if (cents.toString().length === 1) { cents = '0' + cents; }
    return `${dollars}.${cents}`;
  }