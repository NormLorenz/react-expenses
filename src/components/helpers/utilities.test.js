import * as utilities from './utilities';

it('calculateDate', () => {
  expect(utilities.calculateDate('12/12', 2015)).toMatchObject({ date: '2015-12-12T08:00:00.000Z', taxYear: 2015 });
  expect(utilities.calculateDate('6/6', 2015)).toMatchObject({ date: '2015-06-06T07:00:00.000Z', taxYear: 2015 });
  expect(utilities.calculateDate('1/1', 2015)).toMatchObject({ date: '2015-01-01T08:00:00.000Z', taxYear: 2015 });
  expect(utilities.calculateDate('1/1/2010', 2015)).toMatchObject({ date: '2010-01-01T08:00:00.000Z', taxYear: 2010 });
  expect(utilities.calculateDate('2/29', 2015)).toMatchObject({ date: '2015-01-01T08:01:00.000Z', taxYear: 2015 });
  expect(utilities.calculateDate('2/29/1988', 2015)).toMatchObject({ date: '1988-02-29T08:00:00.000Z', taxYear: 1988 });
  expect(utilities.calculateDate('sdf', 2015)).toMatchObject({ date: '2015-01-01T08:01:00.000Z', taxYear: 2015 });
});

it('testDate', () => {
  expect(utilities.testDate()).toEqual('Thu Jan 01 2015 00:01:00 GMT-0800');
});

it('sum', () => {
  expect(utilities.sum(1, 2)).toBe(3);
  expect(utilities.sum(1, 2)).toEqual(3);
  expect(utilities.sum(11, 22)).toEqual(33);
});