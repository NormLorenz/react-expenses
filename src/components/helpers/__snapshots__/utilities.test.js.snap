// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshop 1`] = `
Object {
  "date": "2012-12-12T08:00:00.000Z",
  "taxYear": 2012,
}
`;
